### ISSUE TITLE
Trying to add restart switch fails

### ISSUE THREAD

**Issue by heylers** [esphome/issues#4970]
### The problem

I have a working Bluetooth proxy that I would like to restart periodically.

I am adding

switch:
  - platform: restart
    name: "BT Proxy 1 Restart"   

to the yaml and when I install, I get an error:

ERROR: Could not find a version that satisfies the requirement urllib3<2
ERROR: No matching distribution found for urllib3<2
*** Error 1

and the install fails.

Any help appreciated!

Note:  the full yaml is below and works without the last three lines ....

### Which version of ESPHome has the issue?

2023.9.3

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

n/a but I am on 2023.10

### What platform are you using?

ESP32

### Example YAML snippet

```yaml
esphome:
  name: esphome-web-975018
  friendly_name: Bluetooth Proxy

esp32:
  board: esp32dev
  framework:
    type: esp-idf

# Enable logging
logger:

# Enable Home Assistant API
api:
  encryption:
    key: "....."

ota:
  

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  #ap:
  #  ssid: "Esphome-Web-975018"
  # password: "eHhxYgZjXjSJ"


esp32_ble_tracker:
  scan_parameters:
    interval: 333ms
    window: 333ms
    active: true


bluetooth_proxy:
  active: true


switch:
  - platform: restart
    name: "BT Proxy 1 Restart"
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2023.9.3
INFO Reading configuration /config/esphome-web-975018.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing esphome-web-975018 (board: esp32dev; framework: espidf; platform: platformio/espressif32@5.4.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40405.230623 (4.4.5) 
 - tool-cmake @ 3.16.4 
 - tool-ninja @ 1.7.1 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Installing ESP-IDF's Python dependencies
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f113a810f40>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/urllib3/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f113a810cd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/urllib3/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f113a810ac0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/urllib3/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f113a810850>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/urllib3/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f113a696280>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/urllib3/
ERROR: Could not find a version that satisfies the requirement urllib3<2
ERROR: No matching distribution found for urllib3<2
*** Error 1
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.4
Linking .pioenvs/esphome-web-975018/firmware.elf
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/esphome/components/api/api_connection.o:(.literal._ZN7esphome3api13APIConnection14switch_commandERKNS0_20SwitchCommandRequestE+0x8): undefined reference to `esphome::switch_::Switch::turn_on()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/esphome/components/api/api_connection.o:(.literal._ZN7esphome3api13APIConnection14switch_commandERKNS0_20SwitchCommandRequestE+0xc): undefined reference to `esphome::switch_::Switch::turn_off()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/esphome/components/api/api_connection.o: in function `esphome::api::APIConnection::switch_command(esphome::api::SwitchCommandRequest const&)':
/config/.esphome/build/esphome-web-975018/src/esphome/components/api/api_connection.cpp:485: undefined reference to `esphome::switch_::Switch::turn_on()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /config/.esphome/build/esphome-web-975018/src/esphome/components/api/api_connection.cpp:487: undefined reference to `esphome::switch_::Switch::turn_off()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/main.o:(.literal._Z5setupv+0x88): undefined reference to `vtable for esphome::restart::RestartSwitch'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/main.o:(.literal._Z5setupv+0x8c): undefined reference to `vtable for esphome::restart::RestartSwitch'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/main.o:(.literal._Z5setupv+0xd0): undefined reference to `esphome::switch_::Switch::Switch()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/main.o: in function `setup()':
/config/.esphome/build/esphome-web-975018/src/main.cpp:168: undefined reference to `esphome::switch_::Switch::Switch()'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/esphome/core/controller.o:(.literal._ZN7esphome10Controller16setup_controllerEb+0x8): undefined reference to `esphome::switch_::Switch::add_on_state_callback(std::function<void (bool)>&&)'
/config/.esphome/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pioenvs/esphome-web-975018/src/esphome/core/controller.o: in function `esphome::Controller::setup_controller(bool)':
/config/.esphome/build/esphome-web-975018/src/esphome/core/controller.cpp:35: undefined reference to `esphome::switch_::Switch::add_on_state_callback(std::function<void (bool)>&&)'
collect2: error: ld returned 1 exit status
*** [.pioenvs/esphome-web-975018/firmware.elf] Error 1
========================= [FAILED] Took 23.85 seconds =========================
```

**Comment by heylers** [comment]
Seems like it was a DNS issue that was resolved by changing to 8.8.8.8 .....

**Comment by heylers** [comment]
Seems like it was a DNS issue that was resolved by changing to 8.8.8.8 .....
